wire-runtime-swift/src/main/swift/ProtoCodable/ReadBuffer.swift

```diff
@@ -116,44 +116,8 @@ extension ReadBuffer {
         return UInt64(littleEndian: value)
     }
 
-    /**
-     * Reads a raw varint from the stream. If larger than 32 bits, discard the upper bits.
-     */
-    func readVarint32() throws -> UInt32 {
```

Comment:
It still may be worth keeping this function around but changing its implementation to `return UInt32(truncatingIfNeeded: readVarint64())`. This does two things:
- Ensures that callsites are a bit more self-documenting as to what they expect the bit-width of the varint to be.
- Provides an optimization point should we eventually discover this is a hotspot. Wire's Kotlin has an optimization for this variant of varint (heh). I'm not saying we need to do the optimization eagerly for Swift, but if we inline this into all the callsites then doing the optimization later means going and having to find all those callsites.

We could even mark it as inline or whatever. Is that a thing? I've forgotten all my Swift...

Reply:
You can mark as inline, but my reason to avoid doing so was that it might be a problem when you _want_ Int32, but go through the UInt32 version. 

We would need to write tests to verify that the truncation of eg -5 and -2^31 behave sanely

Reply:
> It still may be worth keeping this function around but changing its implementation to `return UInt32(truncatingIfNeeded: readVarint64())`

Yes I actually considered the same thing.

Reply:
In cases where we know the value is unsigned, such as for encoded message lengths (probably the most common case by a long shot), this is existing implementation is still likely to be slightly faster since there's no need to allocate a UInt64, just to then truncate it.

It would be nice to profile this quickly to verify (even by just running one of the tests 10,000X with both implementations and comparing). If the difference is negligible, then I'm all for just having the one parsing implementation.

If there's a difference between the two implementations, then I think I'm inclined to keep this but maybe rename it to something like `readUnsignedVarint64()` and make sure the documentation is clear on what it can be used for. Then we'd use that method in places like parsing `.lengthEncoded` messages and `readFieldKey()`.

Reply:
+1, will measure it!

Reply:
@efirestone I measured it like this:

```swift
    func testDecodeVarintInt32() throws {
        let data = Foundation.Data(hexEncoded: "08_FBFFFFFFFFFFFFFFFF01")!
        measureMetrics([.wallClockTime], automaticallyStartMeasuring: false) {
            do {
                try test(data: data) { reader in
                    let value = try reader.decode(tag: 1) {
                        self.startMeasuring()
                        let value = try reader.decode(Int32.self, encoding: ProtoIntEncoding.variable)
                        self.stopMeasuring()
                        return value
                    }
                    XCTAssertEqual(value, -5)
                }
            } catch {
                XCTFail("Exception thrown during performance test.")
            }
        }
    }
```

I got a before and after measurements after 1000 repetitions. I did not see any difference between the two at all.

The total runtime of 1000s after was 260s vs the before was 261s, even faster but honestly could be attributed to various random factors.

Feel free to inspect this tiny html page I generated for it.
[wire_encode_measure.zip](https://github.com/square/wire/files/13425441/wire_encode_measure.zip)

Does that cover your concern around performance metrics with this change? Anything you'd like me to look into further?

Reply:
ping @efirestone if you have time to take a look at this!

Reply:
Going to go ahead and merge this and proceed so we can have a release to fix the issue. It seems performance is not affected by this change.